packages/experience-builder-sdk/src/core/editor/designComponentUtils.ts

```diff
@@ -0,0 +1,107 @@
+import { EntityStore } from '@contentful/visual-sdk';
+import {
+  CompositionNode,
+  CompositionDataSource,
+  CompositionUnboundValues,
+  CompositionComponentNode,
+  CompositionComponentPropValue,
+  Composition,
+} from '../../types';
+import { generateRandomId } from '../../utils';
+import { designComponentsRegistry } from '../../blocks/VisualEditorContext';
+
+export const deserializeDesignComponentNode = ({
+  node,
+  parentId,
+  compositionDataSource,
+  compositionUnboundValues,
+}: {
+  node: CompositionNode;
+  parentId?: string;
+  compositionDataSource: CompositionDataSource;
+  compositionUnboundValues: CompositionUnboundValues;
+}): CompositionComponentNode => {
+  const childNodeVariable: Record<string, CompositionComponentPropValue<'DesignValue'>> = {};
+  const dataSource: CompositionDataSource = {};
+  const unboundValues: CompositionUnboundValues = {};
+
+  for (const [variableName, variable] of Object.entries(node.variables)) {
+    if (variable.type === 'DesignValue') {
+      childNodeVariable[variableName] = variable;
+    } else if (variable.type === 'BoundValue') {
+      const [, uuid, ,] = variable.path.split('/');
+
+      dataSource[uuid] = { ...compositionDataSource[uuid] };
+    } else if (variable.type === 'UnboundValue') {
+      const uuid = variable.key;
+      unboundValues[uuid] = compositionUnboundValues[uuid];
+    }
+  }
+
+  const isDesignComponent = node.definitionId?.startsWith('DesignComponent');
+  const nodeId = isDesignComponent ? node.definitionId?.split('-')[1] : generateRandomId(16);
```

Comment:
Why do we need this random generated ID? Don't all nodes have a nodeId already?

Reply:
Because as at the time we fetch design component entry's children in the SDK, they don't have a node ID and need to be deserialized. Hence we generate a node id for them during deserialization (just like we do during node deserialization in the web app)